flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/nn/QuadTree.scala

```diff
@@ -0,0 +1,305 @@
+import org.apache.flink.ml.metrics.distances.DistanceMetric
+
+import scala.collection.mutable.ListBuffer
+import scala.collection.mutable.PriorityQueue
+
+/**
+ * n-dimensional QuadTree data structure; partitions
+ * spatial data for faster queries (e.g. KNN query)
+ * The skeleton of the data structure was initially
+ * based off of the 2D Quadtree found here:
+ * http://www.cs.trinity.edu/~mlewis/CSCI1321-F11/Code/src/util/Quadtree.scala
+ *
+ * Many additional methods were added to the class both for
+ * efficient KNN queries and generalizing to n-dim.
+ *
+ * @param minVec
+ * @param maxVec
+ */
+class QuadTree(minVec:ListBuffer[Double], maxVec:ListBuffer[Double],distMetric:DistanceMetric){
+  var maxPerBox = 20
+
+  class Node(c:ListBuffer[Double],L:ListBuffer[Double], var children:ListBuffer[Node]) {
+
+    var objects = new ListBuffer[Vector]
+
+    /** for testing purposes only; used in QuadTreeSuite.scala
+      *
+      * @return
+      */
+    def getCenterLength(): (ListBuffer[Double], ListBuffer[Double]) = {
+      (c, L)
+    }
+
+    def contains(obj: Vector): Boolean = {
+      overlap(obj, 0.0)
+    }
+
+    /** Tests if obj is within a radius of the node
+      *
+      * @param obj
+      * @param radius
+      * @return
+      */
+    def overlap(obj: Vector, radius: Double): Boolean = {
+      var count = 0
+      for (i <- 0 to obj.size - 1) {
+        if (obj(i) - radius < c(i) + L(i) / 2 && obj(i) + radius > c(i) - L(i) / 2) {
+          count += 1
+        }
+      }
+
+      if (count == obj.size) {
+        return true
+      } else {
+        return false
+      }
+    }
+
+    /** Tests if obj is near a node:  minDist is defined so that every point in the box
+      * has distance to obj greater than minDist
+      * (minDist adopted from "Nearest Neighbors Queries" by N. Roussopoulos et al.)
+      *
+      * @param obj
+      * @param radius
+      * @return
+      */
+    def isNear(obj: Vector, radius: Double): Boolean = {
+      if (minDist(obj) < radius) {
+        true
+      } else {
+        false
+      }
+    }
+
+    def minDist(obj: Vector): Double = {
+      var minDist = 0.0
+      for (i <- 0 to obj.size - 1) {
+        if (obj(i) < c(i) - L(i) / 2) {
+          minDist += math.pow(obj(i) - c(i) + L(i) / 2, 2)
+        } else if (obj(i) > c(i) + L(i) / 2) {
+          minDist += math.pow(obj(i) - c(i) - L(i) / 2, 2)
+        }
+      }
+      return minDist
+    }
+
+    def whichChild(obj:Vector):Int = {
+
+      var count = 0
+      for (i <- 0 to obj.size - 1){
+        if (obj(i) > c(i)) {
+          count += Math.pow(2,i).toInt
+        }
+      }
+      count
+    }
+
+    def makeChildren() {
+      var cBuff = new ListBuffer[ListBuffer[Double]]
+      cBuff += c
+      var Childrennodes = new ListBuffer[Node]
+      val cPart = partitionBox(cBuff,L,L.length)
+      for (i <- cPart.indices){
+        Childrennodes = Childrennodes :+ new Node(cPart(i), L.map(x => x/2.0), null)
+
+      }
```

Comment:
Hi @danielblazevski, sorry for not being response either. We just had our first conference.

You're right that with the current `Vector` interface of FlinkML you cannot do much if you want to apply algebraic operations. The recommended way to do it at the moment is to convert Flink `Vectors` into `BreezeVectors` which give you all the operations you can think of. It is very convenient to convert from and to `BreezeVector`. Simply add `import org.apache.flink.ml.math.Breeze._` to your sources and then you can call `vector.asBreeze` to obtain a `BreezeVector` from a `Vector` and `breezeVector.fromBreeze` to obtain a `Vector` from a `BreezeVector`. Then you can simply perform all these algebraic operations without having to operate on the underlying data structure.

One thing to note is that the conversion is not so costly, because we use the same data representation as breeze for our vectors. Therefore, the object creation is rather cheap. However, it is better to not convert in every step from and to breeze. It's better to convert once to breeze in the beginning and then in the end convert back to Flink. You can take a look at the `SVM` code to see how we used it there.